src/snowflake/snowpark/_internal/analyzer/select_statement.py

```diff
@@ -1273,6 +1273,9 @@ def distinct(self) -> "SelectStatement":
             # has a limit clause to avoid moving distinct in front of limit.
             and (not self.limit_)
             and (not self.offset)
+            # .order_by(col1).select(col2).distinct() cannot be flattened because
+            # SELECT DISTINCT B FROM TABLE ORDER BY A is not valid SQL
+            and (not (self.order_by and self.projection))
```

Comment:
Is the following needed as well? It's in sort and filter.

can_clause_dependent_columns_flatten(
                derive_dependent_columns(col), self.column_states
)

Reply:
Every select has `projection`. Does this rule out all cases?